apps/prairielearn/src/lib/externalGraderCommon.js

```diff
@@ -21,169 +20,91 @@ export function getJobDirectory(jobId) {
- * @param {any} variant
- * @param {any} question
- * @param {any} course
+ * @param {import('./db-types').Submission} submission
+ * @param {import('./db-types').Variant} variant
+ * @param {import('./db-types').Question} question
+ * @param {import('./db-types').Course} course
  */
-export function buildDirectory(dir, submission, variant, question, course, callback) {
+export async function buildDirectory(dir, submission, variant, question, course) {
   const coursePath = getRuntimeDirectoryForCourse(course);
-  async.series(
-    [
-      (callback) => {
-        // Attempt to remove existing directory first
-        fs.remove(dir, () => {
-          // Ignore error for now
-          callback(null);
-        });
-      },
-      (callback) => {
-        fs.mkdirs(dir, (err) => {
-          if (ERR(err, callback)) return;
-          callback(null);
-        });
-      },
-      (callback) => {
-        fs.mkdir(path.join(dir, 'serverFilesCourse'), (err) => {
-          if (ERR(err, callback)) return;
-          callback(null);
-        });
-      },
-      (callback) => {
-        fs.mkdir(path.join(dir, 'tests'), (err) => {
-          if (ERR(err, callback)) return;
-          callback(null);
-        });
-      },
-      (callback) => {
-        fs.mkdir(path.join(dir, 'student'), (err) => {
-          if (ERR(err, callback)) return;
-          callback(null);
-        });
-      },
-      (callback) => {
-        fs.mkdir(path.join(dir, 'data'), (err) => {
-          if (ERR(err, callback)) return;
-          callback(null);
-        });
-      },
-      (callback) => {
-        // Copy all specified files/directories into serverFilesCourse/
-        if (question.external_grading_files) {
-          async.each(
-            question.external_grading_files,
-            (file, callback) => {
-              const src = path.join(coursePath, 'serverFilesCourse', file);
-              const dest = path.join(dir, 'serverFilesCourse', file);
-              fs.copy(src, dest, (err) => {
-                if (ERR(err, callback)) return;
-                callback(null);
-              });
-            },
-            (err) => {
-              if (ERR(err, callback)) return;
-              callback(null);
-            },
-          );
-        } else {
-          callback(null);
-        }
-      },
-      (callback) => {
-        // This is temporary while /grade/shared is deprecated but still supported
-        // TODO remove this when we remove support for /grade/shared
-        const src = path.join(dir, 'serverFilesCourse');
-        const dest = path.join(dir, 'shared');
-        fs.copy(src, dest, (err) => {
-          if (ERR(err, callback)) return;
-          callback(null);
-        });
-      },
-      (callback) => {
-        // Tests might not be specified, only copy them if they exist
-        const testsDir = path.join(coursePath, 'questions', question.directory, 'tests');
-        fs.access(testsDir, fs.constants.R_OK, (err) => {
```

Comment:
I replaced the access+copy with a copy with a check for ENOENT instead. This is safer (avoids TOCTTOU issue). I tested locally and it works as expected.